### ISSUE TITLE
Example question_answering.py - tqdm "ValueError: 1 is not in list"

### ISSUE THREAD

**Issue by ahotrod** [deepset-ai/FARM#208]
Running the current `question_answering.py` example, during preprocessing dataset: 
```
question_answering():
-->     44 data_silo = DataSilo(processor=processor, batch_size=batch_size, distributed=False)
data_silo.py:
        82 self._load_data()
       179 self.data["train"], self.tensor_names = self._get_dataset(train_file)
       149 with tqdm(total=len(dicts), unit=' Dicts', desc="Preprocessing Dataset") as pbar:
-->    150    for dataset, tensor_names in results:
pool.py:
       746 if success:
       747     return value
-->    748 raise value
```
Execution stops with **"ValueError: 1 is not in list"**

I have searched the web for references to this problem with no success.
Anyone experienced this with tqdm and/or have a solution?
Thanks!

**System:**
 - OS: Linux Mint 19.3 based on Ubuntu 18.04. 3 LTS and Linux Kernel 5.0
 - GPU/CPU: 2 x NVIDIA 1080Ti / Intel i7-8700
 - FARM version: 0.3.2
 - Transformers:  2.3.0
 - PyTorch:  1.4.0
 - TensorFlow:  2.1.0
 - Python: 3.7.5
 - tqdm:  4.41.1 (latest version)

**Comment by brandenchan** [comment]
So I have not been able to reproduce the error. But looking at the console output you provided, I suspect it has something to do with multiprocessing. Could you try this again using single processing? To do this you will need to use the latest version of master. Then set DataSilo(...max_processes=1)

**Comment by ahotrod** [comment]
@brandenchan thanks for your prompt response.

Updated FARM:  `pip install --editable .` after `git pull` on master

Modified question_answering.py code:
`data_silo = DataSilo(processor=processor, max_processes=1, batch_size=batch_size, distributed=False)`

After nearly 1/2 hour with only one CPU thread at 100% utilization, execution terminated and received the same error message, "**ValueError: 1 is not in list**".  No progress was indicated on the tqdm scale as before:

`Preprocessing Dataset:   0%|                                                 | 0/19035 [22:36<?, ? Dicts/s]`

Only difference, with the previous, original multiprocessing configuration the stoppage & error message was within seconds, makes sense.  I will try to trace it further.

**Comment by ahotrod** [comment]
Re-visited one of my best-performance Transformers fine-tuning scripts, `xlnet_lrg_sqd2_512_bs48.sh`
to verify performance and all is well:

```
01/23/2020 11:45:37 - INFO - __main__ -   Creating features from dataset file at .
100%|█████████████████████████████████████████████████████████████████████████████████████████████| 442/442 [00:30<00:00, 14.33it/s]
convert squad examples to features: 100%|██████████████████████████████████████████████████| 130319/130319 [06:59<00:00, 310.40it/s]
add example index and unique id: 100%|█████████████████████████████████████████████████| 130319/130319 [00:00<00:00, 1695385.91it/s]
01/23/2020 11:53:11 - INFO - __main__ -   Saving features into cached file ./cached_train_xlnet-large-cased_512
```
tqdm under Transformers is performing as expected.  I'll keep looking.

**Comment by ahotrod** [comment]
Ran another trial with Pytorch==1.3.0 and Tensorflow==2.0.0, gave same error.

**Comment by tholor** [comment]
Hmm.. since we cannot reproduce this issue on our side, let's try to narrow it down together. 
I could think of a few directions:
a) Dataset: Maybe the dataset is corrupt? Are you using the squad data that gets automatically downloaded in the example script?  Can you try to delete the folder "../data/squad20" and let it be downloaded again by the script.

b) FARM version: did you uninstall FARM before installing from master?

c) Can you run any of the other examples (e.g. examples/doc_classification.py)?

d) Can you post all logs for the QA example? 

Hopefully, any of this leads us to a better understanding of the root cause :)

**Comment by ahotrod** [comment]
Created completely new environment from the latest repos:
Python: 3.7.6
PyTorch:  1.4.0
Tensorflow: 2.1.0
Transformers: 2.3.0
Farm:  0.4.0, installed via `pip install --editable .` from new git clone of master

> Hmm.. since we cannot reproduce this issue on our side, let's try to narrow it down together.
> I could think of a few directions:
> a) Dataset: Maybe the dataset is corrupt? Are you using the squad data that gets automatically downloaded in the example script? Can you try to delete the folder "../data/squad20" and let it be downloaded again by the script.
**New download, both dev-v2.0.json & train-v2.0.json are dated 03Jul2019**

> b) FARM version: did you uninstall FARM before installing from master?\
**Completely new environment, same error "ValueError: 1 is not in list".**
> 
> c) Can you run any of the other examples (e.g. examples/doc_classification.py)?
**doc_classification.py works**
> 
> d) Can you post all logs for the QA example?
**Yes, which logs located where?**

**Here's the complete terminal output when run under the new environment listed above:**
[output_term.txt](https://github.com/deepset-ai/FARM/files/4115712/output_term.txt)

> 
> Hopefully, any of this leads us to a better understanding of the root cause :) **Indeed**

**Comment by brandenchan** [comment]
Ok great, thanks. I still haven't been able to replicate the bug but the terminal output gives me another trace as to where it might be. Since you have an editable FARM installed, could you please replace line 358 in data_handler/input_features.py with this code block and then post the terminal output once again?
    ```
    try:
        seq_2_start_t = segment_ids.index(1)
    except ValueError:
        print(sample)
        print(input_ids)
        print(start_of_word)
        print(segment_ids)
```
I feel we're getting closer!

**Comment by brandenchan** [comment]
Just now, we managed to replicate the issue. It is related to huggingface/transformers#2432. Working on a fix now

**Comment by brandenchan** [comment]
#216 has now been merged into master. @ahotrod could you try it and verify that it fixes the problem?

**Comment by ahotrod** [comment]
@brandenchan @tholor thanks!

Yes indeed it works with the new environment as I described in my last post above.  It's a data parallel solution utilizing my hybrid-liquid-cooled GPU 0 and open-framed-cooled GPU1: [qa_nvidia_smi.txt](https://github.com/deepset-ai/FARM/files/4119343/qa_nvidia_smi.txt)

By the fourth Sample, "_What intractability means in practice is open to debate_" there was a stoppage:

`RuntimeError: CUDA out of memory. Tried to allocate 2.72 GiB (GPU 0; 10.92 GiB total capacity; 7.12 GiB already allocated; 2.03 GiB free; 8.36 GiB reserved in total by PyTorch)` [qa_output_term.txt](https://github.com/deepset-ai/FARM/files/4119423/qa_output_term.txt)

but that's hardware dependent and I can deal with that.

Thanks again for your prompt responses and good luck with the FARM.  I am working on a useful, practical closed-domain question answering project, so perhaps I can contribute with feedback of my experiences going forward.

**Comment by tholor** [comment]
Glad that we could nail it down & fix it together, @ahotrod !

QA is also the main task we work on these days and we made already some learnings with large scale deployments in the industry, that will find their way into FARM and our new [haystack](https://github.com/deepset-ai/haystack/) project. 
Would be really nice to exchange on experiences and features that would simplify your dev cycle. You can reach us at hello [at] deepset.ai

**Comment by ahotrod** [comment]
@tholor - same here

FARM, haystack, and your services may be valuable to our project.
Is _hello [at] deepset.ai_ the appropriate email address to send you a project outline & description?

> Regarding the memory issue. Have you considered gradient accumulation? See number 8 here

Yes, very familiar with gradient accumulation.  Rely on it extensively in Transformers, e.g. the shell script mentioned earlier `xlnet_lrg_sqd2_512_bs48.sh` for fine-tuning XLNET with my hardware-challenged setup.   I'm waiting on NVIDIA Ampere releases for a system GPU upgrade.

Unfortunately recent mods & reconfigs to the excellent Transformers libraries may have negatively impacted my results with new XLNET fine-tuning runs, https://github.com/huggingface/transformers/issues/2651 - Hopefully it's only a temporary situation and a fix is not far off.  There's always BERT, ROBERTA & ALBERT for QA alternatives, although XLNET has performed well in our early application testing.

**Comment by tholor** [comment]
> There's always BERT, ROBERTA & ALBERT for QA alternatives, although XLNET has performed well in our early application testing.

Nice! We mostly use BERT & Roberta these days, but I guess we should benchmark again with XLNET for some scenarios. ALBERT is still our big hope, but it seems a bit tricky to get really stable, reproducible results with it. 
 
> Is hello [at] deepset.ai the appropriate email address to send you a project outline & description?

yep
